src/ImageDetector/index.js

```diff
@@ -0,0 +1,162 @@
+import * as tf from '@tensorflow/tfjs';
+import handleArguments from '../utils/handleArguments';
+import { mediaReady } from '../utils/imageUtilities';
+
+/**
+ * @typedef {Object} SpecificDetectorImplementation
+ *
+ * @property {Promise<boolean>} ready - lets the parent detector know that the
+ * specific implementation is ready.
+ *
+ * @property {(input: tf.Tensor3D) => Promise<Array>} detect - core detection method.
+ * the ImageDetector will call the `detect` method of the specific implementation.
+ * It should accept a TensorFlow tensor?? or an image??
+ * And return an array of detections??
+ */
+
+/**
+ * Helper class for handling the public API of detector models (facemesh, etc.)
+ * Exposes the public methods for single and continuous detection.
+ * Executes the detection using whatever model is passed to the constructor.
+ */
+export default class ImageDetector {
+
+  /**
+   * @param {SpecificDetectorImplementation} specificImplementation
+   */
+  constructor(specificImplementation) {
+    /**
+     * @type {SpecificDetectorImplementation}
+     */
+    this.implementation = specificImplementation;
```

Comment:
@gohai @shiffman After reading your comments I think that my terminology here might be over the top.  I wanted to be really clear about what is what.  But it's not going to translate if the person reading it doesn't understand the terminology and it sounds like your students will not be familiar with concepts and terms like interface vs implementation and concretion vs abstraction.  Maybe it is better to keep it simple instead.

```js
  /**
   * @param {DetectorModel} model
   */
  constructor(model) {
    /**
     * @type {DetectorModel}
     */
    this.model = model;
```

I'm also tempted to add a link to https://en.wikipedia.org/wiki/Dependency_injection in one of those JSDoc blocks! 😝

Reply:
I think I would err on the side of keeping things simple for now, maybe we can open an issue for discussion around using "interfaces" in the codebase for later? For now, I would priortize the codebase being as friendly for beginners to contribute to even if it's not making use of the latest and greatest JS features / best practices.